From 39bcd823379bf8a5e1f1a0ef351ac4550f72f389 Mon Sep 17 00:00:00 2001
From: Plato Mavropoulos <platomav@users.noreply.github.com>
Date: Sun, 4 Feb 2018 02:41:35 +0200
Subject: [PATCH 2/4] Fix hrules "ALL" Line Drawing Mode

When using "ALL" hrules, the bottom border now appears and is drawn properly at the last row.
---
 prettytable/__init__.py    |  3 ++-
 prettytable/prettytable.py | 11 ++++++-----
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/prettytable/__init__.py b/prettytable/__init__.py
index 5ce797e..23f2029 100644
--- a/prettytable/__init__.py
+++ b/prettytable/__init__.py
@@ -33,5 +33,6 @@
 __version__ = "0.9.2"
 
 from .prettytable import PrettyTable
-from .prettytable import ALL, BOX_CHARS, HEADER, MSWORD_FRIENDLY, NONE
+from .prettytable import FRAME, ALL, NONE, HEADER
+from .prettytable import DEFAULT, MSWORD_FRIENDLY, PLAIN_COLUMNS, RANDOM, BOX_CHARS
 from .factory import from_csv, from_db_cursor, from_html, from_html_one
diff --git a/prettytable/prettytable.py b/prettytable/prettytable.py
index 3417f8a..50eda4b 100644
--- a/prettytable/prettytable.py
+++ b/prettytable/prettytable.py
@@ -1439,11 +1439,12 @@ class PrettyTable(object):
             lines.append(self._hrule)
 
         # Add rows
-        for row in formatted_rows:
-            lines.append(self._stringify_row(row, options))
+        for row in range(0, len(formatted_rows)):
+            if row == len(formatted_rows) - 1 : lines.append(self._stringify_row(formatted_rows[row], options, True))
+            else : lines.append(self._stringify_row(formatted_rows[row], options, False))
 
         # Add bottom of border
-        if options["border"] and options["hrules"] == FRAME:
+        if options["border"] and options["hrules"] in (ALL, FRAME):
             lines.append(self._stringify_bottom_hrule(options))
 
         return self._unicode("\n").join(lines)
@@ -1648,7 +1649,7 @@ class PrettyTable(object):
             bits.append(self._hrule)
         return "".join(bits)
 
-    def _stringify_row(self, row, options):
+    def _stringify_row(self, row, options, isLastRow):
 
         for index, field, value, width, in zip(range(0, len(row)), self._field_names, row, self._widths):
             # Enforce max widths
@@ -1711,7 +1712,7 @@ class PrettyTable(object):
                 bits[y].pop()
                 bits[y].append(options["border_vertical_char"])
 
-        if options["border"] and options["hrules"] == ALL:
+        if options["border"] and not isLastRow and options["hrules"] == ALL:
             bits[row_height - 1].append("\n")
             bits[row_height - 1].append(self._hrule)
 
-- 
2.10.5

